tokenizers/src/utils/from_pretrained.rs

```diff
@@ -0,0 +1,139 @@
+///
+/// If the `TOKENIZERS_CACHE` environment variable is set, we just return it. It is the
+/// responsibility of the user to make sure this path is correct.
+///
+/// Otherwise, we try to use the default cache directory as defined for each OS:
+///     - macOS: `/Users/{user}/Library/Caches/huggingface/tokenizers`
+///     - linux: `/home/{user}/.cache/huggingface/tokenizers`
+///     - windows: `C:\Users\{user}\AppData\Local\huggingface\tokenizers`
+/// If the default cache directory cannot be found (if the user HOME folder is not defined),
+/// then we fall back on a temporary directory
+fn cache_dir() -> PathBuf {
+    if let Ok(path) = std::env::var("TOKENIZERS_CACHE") {
+        PathBuf::from(path)
+    } else {
+        let mut dir = dirs::cache_dir().unwrap_or_else(std::env::temp_dir);
+        dir.push("huggingface");
+        dir.push("tokenizers");
+        dir
+    }
+}
+
+/// Returns a directory to be used as cache, creating it if it doesn't exist
+///
+/// Cf `cache_dir()` to understand how the cache dir is selected.
+fn ensure_cache_dir() -> std::io::Result<PathBuf> {
+    let dir = cache_dir();
+    std::fs::create_dir_all(&dir)?;
+    Ok(dir)
+}
+
+/// Sanitize a key or value to be used inside the user_agent
+/// The user_agent uses `/` and `;` to format the key-values, so we
+/// replace them by `-`
+fn sanitize_user_agent(item: &str) -> Cow<str> {
+    let mut sanitized = Cow::Borrowed(item);
+    if sanitized.contains('/') {
+        sanitized = Cow::Owned(sanitized.replace('/', "-"));
+    }
+    if sanitized.contains(';') {
+        sanitized = Cow::Owned(sanitized.replace(';', "-"));
+    }
+    sanitized
+}
+
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+// We allow unstable name collisions in this case because we don't care if it
+// starts using the new stable feature when it will be stable. This feature is
+// supposed to be a copy of the one we use anyway.
+// cf https://github.com/rust-lang/rust/issues/79524
+#[allow(unstable_name_collisions)]
+fn user_agent(additional_info: HashMap<String, String>) -> String {
+    let additional_str: String = additional_info
+        .iter()
+        .map(|(k, v)| format!("{}/{}", sanitize_user_agent(k), sanitize_user_agent(v)))
+        .intersperse("; ".to_string())
+        .collect();
+
+    let user_agent = format!(
+        "tokenizers/{}{}",
+        VERSION,
+        if !additional_str.is_empty() {
+            format!("; {}", additional_str)
+        } else {
+            String::new()
+        }
+    );
+
+    user_agent
+}
+
+/// Defines the aditional parameters available for the `from_pretrained` function
+#[derive(Debug, Clone)]
+pub struct FromPretrainedParameters {
+    pub revision: String,
+    pub user_agent: HashMap<String, String>,
+    pub auth_token: Option<String>,
+}
+
+impl Default for FromPretrainedParameters {
+    fn default() -> Self {
+        Self {
+            revision: "main".into(),
+            user_agent: HashMap::new(),
```

Comment:
Shouldn't this be "tokenizers", "rust" or something by default ?

Reply:
It is, by default the user agent always contains `tokenizers/{RUST_VERSION}`, this is for additional items